File: 0739-daily-temperatures/0739-daily-temperatures.py
class Solution:
    def dailyTemperatures(self, t: List[int]) -> List[int]:
        ans = [0 for _ in t]
        stack = [] # [ [temp, idx] entries ]

        for idx, temp in enumerate(t):
            while stack and temp > stack[-1][0]:
                curr_temp, curr_idx = stack.pop()
                ans[curr_idx] = idx - curr_idx
            stack.append([temp, idx])
        
        return ans

File: 0739-daily-temperatures/0739-daily-temperatures.cpp
class Solution {
public:
    vector<int> dailyTemperatures(vector<int>& temperatures) {
        int temp = 0;
        int n = temperatures.size();
        vector<int> ans(n, 0);
        for(int i=n-1;i>=0;i--) {
            if(temperatures[i] >= temp) {
                temp = temperatures[i];
            } else {
                int j = i+1;
                while(temperatures[j] <= temperatures[i]) {
                    j += ans[j];
                }
                ans[i] = j - i;
            }
        }
        return ans;
    }
};